turbine/src/quic_endpoint.rs

```diff
@@ -0,0 +1,429 @@
+            RwLock,
+        },
+    },
+};
+
+const CLIENT_CHANNEL_CAPACITY: usize = 1 << 20;
+const INITIAL_MAX_UDP_PAYLOAD_SIZE: u16 = 1280;
+const ALPN_TURBINE_PROTOCOL_ID: &[u8] = b"solana-turbine";
+// Empty server_name will error ConnectError(InvalidDnsName("")).
+const CONNECT_SERVER_NAME: &str = "solana-turbine";
+
+const CONNECTION_CLOSE_ERROR_CODE_DROPPED: VarInt = VarInt::from_u32(1);
+const CONNECTION_CLOSE_ERROR_CODE_INVALID_IDENTITY: VarInt = VarInt::from_u32(2);
+const CONNECTION_CLOSE_ERROR_CODE_REPLACED: VarInt = VarInt::from_u32(3);
+
+const CONNECTION_CLOSE_REASON_DROPPED: &[u8] = b"DROPPED";
+const CONNECTION_CLOSE_REASON_INVALID_IDENTITY: &[u8] = b"INVALID_IDENTITY";
+const CONNECTION_CLOSE_REASON_REPLACED: &[u8] = b"REPLACED";
+
+type ConnectionCache = HashMap<(SocketAddr, Option<Pubkey>), Arc<RwLock<Option<Connection>>>>;
+
+#[derive(Error, Debug)]
+pub enum Error {
+    #[error(transparent)]
+    CertificateError(#[from] RcgenError),
+    #[error(transparent)]
+    ConnectError(#[from] ConnectError),
+    #[error(transparent)]
+    ConnectionError(#[from] ConnectionError),
+    #[error("Channel Send Error")]
+    ChannelSendError,
+    #[error("Invalid Identity: {0:?}")]
+    InvalidIdentity(SocketAddr),
+    #[error(transparent)]
+    IoError(#[from] IoError),
+    #[error(transparent)]
+    SendDatagramError(#[from] SendDatagramError),
+    #[error(transparent)]
+    TlsError(#[from] rustls::Error),
+}
+
+#[allow(clippy::type_complexity)]
+pub fn new_quic_endpoint(
+    runtime: Arc<Runtime>,
+    keypair: &Keypair,
+    socket: UdpSocket,
+    address: IpAddr,
+    sender: Sender<(Pubkey, SocketAddr, Bytes)>,
+) -> Result<(Endpoint, AsyncSender<(SocketAddr, Bytes)>, JoinHandle<()>), Error> {
+    let (cert, key) = new_self_signed_tls_certificate(keypair, address)?;
+    let server_config = new_server_config(cert.clone(), key.clone())?;
+    let client_config = new_client_config(cert, key)?;
+    let mut endpoint = {
+        let _guard = runtime.enter();
+        Endpoint::new(
+            EndpointConfig::default(),
+            Some(server_config),
+            socket,
+            TokioRuntime,
+        )?
+    };
+    endpoint.set_default_client_config(client_config);
+    let cache = Arc::<RwLock<ConnectionCache>>::default();
+    let (client_sender, client_receiver) = tokio::sync::mpsc::channel(CLIENT_CHANNEL_CAPACITY);
+    let server_task = runtime.spawn(run_server(endpoint.clone(), sender.clone(), cache.clone()));
+    let client_task = runtime.spawn(run_client(endpoint.clone(), client_receiver, sender, cache));
+    let run_endpoint = move || {
+        let task = futures::future::try_join(server_task, client_task);
+        if let Err(err) = runtime.block_on(task) {
+            error!("runtime.block_on(task): {err:?}");
+        }
+    };
+    let join_handle = std::thread::Builder::new()
+        .name(String::from("solTurbineQuic"))
+        .spawn(run_endpoint)?;
```

Comment:
I know this is how it's done for the TPU, but is there a reason to spawn a thread to block on the task? The runtime is already doing the work after you call `runtime.spawn`, so it would be simpler to return the async task, so then you can just have the common runtime block on all of the tasks once they're all started.

On the flip-side, if there are plans to abandon the async runtime, it's probably easier to keep a thread around.